mlflow/store/_unity_catalog/registry/rest_store.py

```diff
@@ -675,12 +676,15 @@ def _local_model_dir(self, source, local_model_path):
                     f"the source artifact location exists and that you can download from "
                     f"it via mlflow.artifacts.download_artifacts()"
                 ) from e
-            # Clean up temporary model directory at end of block. We assume a temporary
-            # model directory was created if the `source` is not a local path (must be downloaded
-            # from remote to a temporary directory)
-            yield local_model_dir
-            if not os.path.exists(source):
-                shutil.rmtree(local_model_dir)
+            try:
+                yield local_model_dir
+            finally:
+                # Clean up temporary model directory at end of block. We assume a temporary
+                # model directory was created if the `source` is not a local path
+                # (must be downloaded from remote to a temporary directory) and
+                # `local_model_dir` is not a UC volumes path
+                if not os.path.exists(source) and not _is_uc_volumes_path(local_model_dir):
+                    shutil.rmtree(local_model_dir)
```

Comment:
When `log_model` is called with `registered_model_name`, we don't reach here as `_register_model` is called `local_path` (!= `None`):

https://github.com/mlflow/mlflow/blob/1892d825e837a234a0bd6602066ced99ca515a54/mlflow/models/model.py#L743-L749

Reply:
Good catch on this and thanks for the additional testing!

Reply:
Can we mock the control logic flow here to emulate the process just to protect against future modification to the logic in #12545 and this PR from accidentally causing a regression? (this is potentially a scary regression to create as it could permanently delete a stored model instance)

Reply:
I'd keep the regex! We could consider canonicalizing the volume path (e.g. `/volume/` -> `/Volumes/`).